### PR TITLE
fix(@desktop/profile): Contact requests should support a message

### LINKED ISSUE #5846: Contact requests should support a message
### Description

Contact requests should support sending/receiving a message.

### Acceptance Criteria

- [x] profile -> messaging ->  contacts, requests, and blocked users -> send contact request to chat. User should be able to type a message to send as part of the contact request
- [x] user receiving contact request should see the message

### PR DESCRIPTION
**Pull request by MishkaRogachev**
Close #5846

### What does the PR do

Change send, accept and dismiss contact request implementations

### Affected areas

Profile, Chat

### Screenshot of functionality

image

https://user-images.githubusercontent.com/2522130/172174346-13131b7c-6451-40b2-bfb3-d820eced537d.mov

### CODE REVIEW

File: src/app_service/service/contacts/service.nim [hunk 1]
```diff
@@ -301,6 +301,7 @@ QtObject:
         let msg = response.error.message
         error "error adding contact ", msg
         return
+      # FIXME: here i want to send message
```
Comment (MishkaRogachev) [inline src/app_service/service/contacts/service.nim:304]: @saledjenic Here I should send a message, but in the current approach, there is no difference between just adding and requesting contact. Should I separate these implementations?

Comment (saledjenic) [reply src/app_service/service/contacts/service.nim:304]: @MishkaRogachev my understanding is... there are 2 options for sending a contact request to anybody (but that's maybe more a question for @John-44):
1. from contact's profile popup, clicking on "Add to contact" button, in this case you're just sending a contact request, without identity verification request. If you want to send identity verification request later you should go to `profile -> messaging -> contacts` find that contact in appropriate section there and send identity verification request.
2. clicking on "Send contact request to chat key" button from `profile -> messaging -> contacts` section, which opens a search contact popup with a field to enter a message and that message will be the first step in identity verification process for contact you're adding. So in this second case sending a contact request combines two things, adding contact + running identity verification process.

Comment (saledjenic) [reply src/app_service/service/contacts/service.nim:304]: @MishkaRogachev for running identity verification process you should firstly check what is the state on the status-go side of that feature, cause you will need that for this ticket.

Comment (MishkaRogachev) [reply src/app_service/service/contacts/service.nim:304]: Thanks, so probably should add sending message after identity verification process?

Comment (saledjenic) [reply src/app_service/service/contacts/service.nim:304]: @MishkaRogachev as we discussed on the last UI meeting on Friday, that message you're able to write when you're adding somebody to your contacts doesn't have anything with the "Identity Verification" process, don't look at it that way. That message is just here to "say hello..." to a contact you're adding. But now I am thinking maybe we should add that message as the first message in `1:1` chat with the contact you're adding once he accepts your contact request. Will ask that now in a `desktop-ui` channel.

Comment (MishkaRogachev) [reply src/app_service/service/contacts/service.nim:304]: @saledjenic thank you for helping. I've asked about this mobile team and they have already implemented this in the status go with mutual contacts feature: https://github.com/status-im/status-go/pull/2665/files

Comment (saledjenic) [reply src/app_service/service/contacts/service.nim:304]: @MishkaRogachev then I am not sure if that is what we want to achieve. It's more a question for @John-44 now. :)

File: src/app_service/service/contacts/service.nim [hunk 2]
```diff
@@ -296,15 +296,35 @@ QtObject:
...
+        error "error sending contact request", msg
+        return
+      self.saveContact(contact)
+      self.events.emit(SIGNAL_CONTACT_ADDED, ContactArgs(contactId: contact.id))
+    except Exception as e:
+      error "an error occurred while sending contact request", msg=e.msg
+
+  proc acceptContactRequest*(self: Service, publicKey: string) =
+    try:
+      # TODO: does we need this logic on both (sendContactRequest and acceptContactRequest) sides?
```
Comment (MishkaRogachev) [inline src/app_service/service/contacts/service.nim:312]: @saledjenic should i save (and refactor to other method) this logic for both cases?

File: src/backend/contacts.nim [hunk 3]
```diff
@@ -21,25 +21,31 @@ proc unblockContact*(id: string): RpcResponse[JsonNode] {.raises: [Exception].}
...
+proc sendContactRequest*(id: string, message: string): RpcResponse[JsonNode] {.raises: [Exception].} =
   let payload = %* [{
       "id": id,
-      "ensName": ensName
+      "message": message
     }]
-  result = callPrivateRPC("addContact".prefix, payload)
+  result = callPrivateRPC("sendContactRequest".prefix, payload)
+
+proc acceptContactRequest*(id: string): RpcResponse[JsonNode] {.raises: [Exception].} =
```
Comment (MishkaRogachev) [inline src/backend/contacts.nim:38]: @saledjenic Here is an another problem. `acceptContactRequest` need to pass the id of the contact request, not of contact. But we don't have entity for contact request. What sould i do, add it to contact as optional field?

File: src/app_service/service/contacts/service.nim (hunk 2 above)
Comment (saledjenic) [reply src/app_service/service/contacts/service.nim:312]: @MishkaRogachev as we discussed in private fill free to update logic here according to the new changes introduced on the status-go side.

File: src/backend/contacts.nim (hunk 3 above)
Comment (saledjenic) [reply src/backend/contacts.nim:38]: @MishkaRogachev feel free to add new hash map which will map appropriate contact id to appropriate contact request id, that way you will know when you accept somebody's contact request what contact request id is appropriate one for his contact id. But also pay attention if somebody doesn't accept contact request once he receives it, to populate that map correctly on the next app run. That means you need to check do we have a new rpc endpoint which will return currently active contact requests ids for logged in users or maybe that is added as part of fetch all contacts call we're doing on the app start.

Comment (MishkaRogachev) [reply src/backend/contacts.nim:38]: @michaelsbradleyjr here i ned to store map contact.id-contact_request.id. How can I promoute contact request signal from activity center to conatcts/service.nim?

Comment (saledjenic) [reply src/backend/contacts.nim:38]: @MishkaRogachev why do you need to get anything from activity center? AC doesn't need to determine anything in the app, it stores only notifications which are generated by logic beyond AC. So if you need the list of contact requests then you should check do we receive that list as a part of fetch all contacts or is there any special endpoint where you can get the list of pending contact requests.

File: src/app_service/service/activity_center/service.nim [hunk 4]
```diff
@@ -187,6 +191,10 @@ QtObject:
       self.events.emit(chat_service.SIGNAL_CHAT_UPDATE,
         ChatUpdateArgsNew(messages: messages, chats: chats))
 
+      #if (response.notificationType == ActivityCenterNotificationType.ContactRequest)
```
[the remark this answers is not in the document]
Comment (saledjenic) [reply src/app_service/service/activity_center/service.nim]: @MishkaRogachev what do you need here?

File: src/backend/contacts.nim (hunk 3 above)
Comment (endulab) [reply src/backend/contacts.nim:38]: @MishkaRogachev , AC is a separate thing. It also should receive information about contact requests, but only to show them in AC list. I see that contact_service is another endpoint which receives contact requests and probably this is a part you should focus on.

File: src/app_service/service/activity_center/service.nim (hunk 4 above)
[the remark this answers is not in the document]
Comment (MishkaRogachev) [reply src/app_service/service/activity_center/service.nim]: upd: fixed

**Review by saledjenic — APPROVED** [review]

**Comment by elina2015** [comment]
@MishkaRogachev hey, I have some found issues. Could you kindly check them?
1. User A send a contact request without a message to contact B. 
    User B don't see the request and the user A don't see the request in Pending
2. The counter of characters is shown under the field of public chat key, but it should be under the message field:
image
3. The fields are shown very close to the form edge:
image

File: src/app_service/service/contacts/service.nim [hunk 5]
```diff
@@ -297,25 +297,47 @@ QtObject:
...
+        let msg = response.error.message
+        error "error accepting contact request", msg
+        return
+
+      # TODO: does we need this logic on both (sendContactRequest and acceptContactRequest) sides?
+      var contact = self.getContactById(publicKey)
+      if not contact.added:
+        contact.added = true
+      else:
+        contact.blocked = false
```
Comment (endulab) [inline src/app_service/service/contacts/service.nim:324]: If you accept a request from a contact and it is added, it gets blocked ?
What is the application use case ?

File: src/app_service/service/contacts/service.nim [hunk 6]
```diff
@@ -297,25 +297,47 @@ QtObject:
...
+  proc acceptContactRequest*(self: Service, publicKey: string) =
+    try:
+      # NOTE: publicKey used for accepting last request
+      let response = status_contacts.acceptLatestContactRequestForContact(publicKey)
+      if(not response.error.isNil):
+        let msg = response.error.message
+        error "error accepting contact request", msg
+        return
+
+      # TODO: does we need this logic on both (sendContactRequest and acceptContactRequest) sides?
```
Comment (endulab) [inline src/app_service/service/contacts/service.nim:319]: What about this TODO ?

**Review by endulab — CHANGES_REQUESTED** [review]

File: src/app_service/service/contacts/service.nim (hunk 5 above)
Comment (MishkaRogachev) [reply src/app_service/service/contacts/service.nim:324]: I've saved it from older implementation, probably @saledjenic knows do we still need this logic. But it works different: if you add blocked contact - unblock it

File: src/app_service/service/contacts/service.nim (hunk 6 above)
Comment (MishkaRogachev) [reply src/app_service/service/contacts/service.nim:319]: This TODO is for farther contact request implementation but mostly depend on yours first question

File: src/app_service/service/contacts/service.nim (hunk 5 above)
Comment (endulab) [reply src/app_service/service/contacts/service.nim:324]: Ahh, I see. Sounds ok.

**Comment by MishkaRogachev** [comment]
> @MishkaRogachev hey, I found some issues. Could you kindly check them?
> 
> 1. User A sends a contact request without a message to contact B.
>    User B doesn't see the request and the user A don't see the request in Pending
> 2. The counter of characters is shown very close to the field of public chat key:
> 
> image
> 
> 3. The fields are shown very close to the form edge:
> image

1 Fixed with limiting sending empty contact request
2-3 Fixed:
Screenshot 2022-06-15 at 14 10 22

**Review by endulab — APPROVED** [review]

**Pull request merged**